crates/common/src/account.rs

```diff
@@ -0,0 +1,163 @@
+use anyhow::{anyhow, Result};
+use prism_keys::{Signature, SigningKey, VerifyingKey};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    operation::{Operation, ServiceChallenge},
+    transaction::Transaction,
+};
+
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Hash, Default)]
+/// Represents an account or service on prism, making up the values of our state
+/// tree.
+pub struct Account {
+    /// The unique identifier for the account.
+    pub id: String,
+
+    /// The transaction nonce for the account.
+    pub nonce: u64,
+
+    /// The current set of valid keys for the account. Any of these keys can be
+    /// used to sign transactions.
+    pub valid_keys: Vec<VerifyingKey>,
+
+    /// Arbitrary signed data associated with the account, used for bookkeeping
+    /// externally signed data from keys that don't live on Prism.
+    pub signed_data: Vec<(VerifyingKey, Vec<u8>)>,
+
+    /// The service challenge for the account, if it is a service.
+    pub service_challenge: Option<ServiceChallenge>,
+}
+
+impl Account {
+    /// Creates a [`Transaction`] that can be used to update or create the
+    /// account. The transaction produced could be invalid, and will be
+    /// validated before being processed.
+    pub fn prepare_transaction(
+        &self,
+        account_id: String,
+        operation: Operation,
+        sk: &SigningKey,
+    ) -> Result<Transaction> {
+        let vk = sk.verifying_key();
+
+        let mut tx = Transaction {
+            id: account_id,
+            nonce: self.nonce,
+            operation,
+            signature: Signature::Placeholder,
+            vk,
+        };
+
+        tx.sign(sk)?;
+
+        Ok(tx)
+    }
+
+    /// Validates and processes an incoming [`Transaction`], updating the account state.
+    pub fn process_transaction(&mut self, tx: &Transaction) -> Result<()> {
+        self.validate_transaction(tx)?;
+        self.process_operation(&tx.operation)?;
+        self.nonce += 1;
+        Ok(())
+    }
+
+    /// Validates a transaction against the current account state.
+    pub fn validate_transaction(&self, tx: &Transaction) -> Result<()> {
+        if tx.nonce != self.nonce {
+            return Err(anyhow!(
+                "Nonce does not match. {} != {}",
+                tx.nonce,
+                self.nonce
+            ));
+        }
+
+        match tx.operation {
+            Operation::CreateAccount { .. } | Operation::RegisterService { .. } => {}
+            _ => {
+                if tx.id != self.id {
+                    return Err(anyhow!("Transaction ID does not match account ID"));
+                }
+                if !self.valid_keys.contains(&tx.vk) {
+                    return Err(anyhow!("Invalid key"));
+                }
+            }
+        }
+
+        let msg = tx.get_signature_payload()?;
+        tx.vk.verify_signature(&msg, &tx.signature)?;
+
+        self.validate_operation(&tx.operation)?;
```

Comment:
We do this inside of process_operation

Reply:
We can simply not save the account after processing instead of calling validate explicitly -> maybe we can make validate a private method

Reply:
done